File: leetcode-cn/116.填充每个节点的下一个右侧节点指针.py
#
# @lc app=leetcode.cn id=116 lang=python3
#
# [116] 填充每个节点的下一个右侧节点指针
#

# @lc code=start


# Definition for a Node.
class Node:
    def __init__(self,
                 val: int = 0,
                 left: 'Node' = None,
                 right: 'Node' = None,
                 next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next


class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return root

        queue = [root]

        while len(queue) != 0:
            num = len(queue)
            for i in range(num):
                current_node = queue.pop(0)

                if i < num - 1:
                    current_node.next = queue[0]

                if current_node.left:
                    queue.append(current_node.left)

                if current_node.right:
                    queue.append(current_node.right)

        return root


# @lc code=end

File: leetcode-cn/C++/116.填充每个节点的下一个右侧节点指针.cpp
/*
 * @lc app=leetcode.cn id=116 lang=cpp
 *
 * [116] 填充每个节点的下一个右侧节点指针
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        queue<Node*> q;

        if(root == nullptr){
            return root;
        }

        q.push(root);

        while(!q.empty()){
            int sz = q.size();
            Node* pre = nullptr;
            for(int i = 0; i < sz; i++){
                Node* cur = q.front();
                q.pop();

                if(cur->left){
                    q.push(cur->left);
                }

                if(cur->right){
                    q.push(cur->right);
                }

                if (pre){
                    pre->next = cur;
                }
                pre = cur;
            }
        }

        return root;
        
    }
};
// @lc code=end

